src/bloqade/gemini/analysis/logical_validation/impls.py

```diff
@@ -0,0 +1,64 @@
+from kirin import interp as _interp
+from kirin.analysis import ForwardFrame, const
+from kirin.dialects import scf
+
+from bloqade.squin import qubit
+from bloqade.validation.analysis.lattice import Error
+
+from .analysis import GeminiLogicalValidationAnalysis
+
+
+@qubit.dialect.register(key="gemini.validate.logical")
+class __QubitGeminiLogicalValidation(_interp.MethodTable):
+
+    @_interp.impl(qubit.New)
+    def new(
+        self,
+        interp: GeminiLogicalValidationAnalysis,
+        frame: ForwardFrame,
+        stmt: qubit.New,
+    ):
+        # TODO: this is actually tricky, since qalloc calls qubit.new multiple times and we have to make sure qalloc is only called once
+        # but it can technically contain many qubit.new calls
+        # if interp.has_allocated_qubits:
+        #     raise ir.ValidationError(
+        #         stmt, "Can only allocate qubits once in a logical Gemini program!"
+        #     )
+
+        # interp.has_allocated_qubits = True
```

Comment:
Actually what matters is if qubit are allocated after the first gate so maybe add a flag for if there ways already a gate operation

Reply:
Actually I'm not even sure if its a problem if there are allocations after the first gate.